pkg/operator/ceph/config/store.go

```diff
@@ -0,0 +1,151 @@
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package config allows a ceph config file to be stored in Kubernetes and mounted as volumes into
+// Ceph daemon containers.
+package config
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/rook/rook/pkg/clusterd"
+	cephconfig "github.com/rook/rook/pkg/daemon/ceph/config"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	storeName = "rook-ceph-config"
+
+	configVolumeName = "rook-ceph-config"
+
+	confFileName = "ceph.conf"
+)
+
+// Store manages storage of the Ceph config file shared by all daemons (if applicable) as well as an
+// updated 'mon_host' which can be mapped to daemon containers and referenced in daemon command line
+// arguments.
+type Store struct {
+	configMapStore *k8sutil.ConfigMapKVStore
+	context        *clusterd.Context
+}
+
+// GetStore returns the Store for the cluster.
+func GetStore(context *clusterd.Context, namespace string, ownerRef *metav1.OwnerReference) *Store {
+	return &Store{
+		configMapStore: k8sutil.NewConfigMapKVStore(namespace, context.Clientset, *ownerRef),
+		context:        context,
+	}
+}
+
+// CreateOrUpdate creates or updates the stored Ceph config based on the cluster info.
+func (s *Store) CreateOrUpdate(clusterInfo *cephconfig.ClusterInfo) error {
+	c := DefaultCentralizedConfigs(s.context.LogLevel)
+
+	/* TODO: in the future merge config overrides from the cluster CRD */
+
+	t, err := c.FileText()
+	if err != nil {
+		return fmt.Errorf("failed to store the Ceph config file. %+v", err)
+	}
+	if err := s.configMapStore.SetValue(storeName, confFileName, t); err != nil {
+		return fmt.Errorf("failed to store the Ceph config file. %+v", err)
+	}
+	logger.Infof("Generated config file:\n%s", t)
+
+	hosts := make([]string, len(clusterInfo.Monitors))
+	members := make([]string, len(clusterInfo.Monitors))
+	i := 0
+	for _, m := range clusterInfo.Monitors {
+		hosts[i] = m.Endpoint
+		members[i] = m.Name
+		i++
+	}
+	// mon_host is string list of mon endpoints separated in form "IP:PORT,IP:PORT,etc."
+	err = s.configMapStore.SetValue(storeName, "mon_host", strings.Join(hosts, ","))
+	if err != nil {
+		return fmt.Errorf("failed to store the Ceph mon_host info. %+v", err)
+	}
+
+	// mon_initial_members is string list of mon ids separated in form "ID,ID,ID,etc."
+	err = s.configMapStore.SetValue(storeName, "mon_initial_members", strings.Join(members, ","))
+	if err != nil {
+		return fmt.Errorf("failed to store the Ceph mon_initial_members info. %+v", err)
+	}
+
+	/*
+		TODO:
+		if Luminous {
```

Comment:
for now we're just storing everything as if luminous, then will add the mimic improvement separately?

Reply:
As discussed in slack, I think we can document how users can switch from the current ceph.conf override to the new way of doing things. I don't think we need to automate that for users since it should be straight forward to convert to it manually as a one-time action at time of upgrade.

Reply:
Right. I wonder if this item might not have been very clear, so I tried to make it a little better. The pseudocode here is intended to be a placeholder for the logic of when to store a config file for the daemons (luminous) and when to use the mon central config for daemons (mimic+).